File: load-test-framework/src/main/java/com/google/pubsub/clients/grpc/CPSSubscriberTask.java
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////
package com.google.pubsub.clients.grpc;

import static com.google.pubsub.flic.controllers.Client.maxOutstandingRequests;

import com.beust.jcommander.JCommander;
import com.google.auth.oauth2.GoogleCredentials;
import com.google.common.base.Function;
import com.google.common.collect.ImmutableList;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.pubsub.clients.common.LoadTestRunner;
import com.google.pubsub.clients.common.MetricsHandler;
import com.google.pubsub.clients.common.Task;
import com.google.pubsub.flic.common.LoadtestProto.StartRequest;
import com.google.pubsub.v1.AcknowledgeRequest;
import com.google.pubsub.v1.PullRequest;
import com.google.pubsub.v1.PullResponse;
import com.google.pubsub.v1.ReceivedMessage;
import com.google.pubsub.v1.SubscriberGrpc;
import io.grpc.Channel;
import io.grpc.ClientInterceptors;
import io.grpc.auth.ClientAuthInterceptor;
import io.grpc.netty.GrpcSslContexts;
import io.grpc.netty.NegotiationType;
import io.grpc.netty.NettyChannelBuilder;
import java.util.Map;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.net.ssl.SSLException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

/**
 * Runs a task that consumes messages from a Cloud Pub/Sub subscription.
 */
class CPSSubscriberTask extends Task {
  private static final Logger log = LoggerFactory.getLogger(CPSSubscriberTask.class);
  private final String subscription;
  private final int pullSize;
  private GoogleCredentials credentials = null;
  private SubscriberGrpc.SubscriberFutureStub[] stubs;
  private int currentStubIdx;

  private CPSSubscriberTask(StartRequest request) {
    super(request, "grpc", MetricsHandler.MetricName.END_TO_END_LATENCY);
    this.subscription = "projects/"
        + request.getProject()
        + "/subscriptions/"
        + request.getSubscription();
    this.pullSize = request.getMaxMessagesPerPull();
    try {
      this.credentials =
          GoogleCredentials.getApplicationDefault()
              .createScoped(ImmutableList.of("https://www.googleapis.com/auth/cloud-platform"));
      stubs = new SubscriberGrpc.SubscriberFutureStub[request.getMaxOutstandingRequests()];
      for (int i = 0; i < stubs.length; i++) {
        stubs[i] = SubscriberGrpc.newFutureStub(getChannel());
      }
    } catch (IOException e) {
      log.error("Unable to get credentials or create channel.", e);
    }
  }

  public static void main(String[] args) throws Exception {
    LoadTestRunner.Options options = new LoadTestRunner.Options();
    new JCommander(options, args);
    LoadTestRunner.run(options, CPSSubscriberTask::new);
  }

  private Channel getChannel() throws SSLException {
    return ClientInterceptors.intercept(
        NettyChannelBuilder.forAddress("pubsub.googleapis.com", 443)
            .maxMessageSize(20000000) // 20 MB
            .sslContext(GrpcSslContexts.forClient().ciphers(null).build())
            .negotiationType(NegotiationType.TLS)
            .build(),
        new ClientAuthInterceptor(credentials, Executors.newSingleThreadExecutor()));
  }

  @Override
  public ListenableFuture<RunResult> doRun() {
    SubscriberGrpc.SubscriberFutureStub stub = getStub();
    ListenableFuture<PullResponse> responseFuture = stub.pull(
        PullRequest.newBuilder()
            .setSubscription(subscription)
            .setMaxMessages(pullSize)
            .build());
    Function<PullResponse, RunResult> callback = (response) -> {
      RunResult out = new RunResult();
      List<String> ackIds = new ArrayList<>(response.getReceivedMessagesCount());
      List<ReceivedMessage> receivedList = response.getReceivedMessagesList();
      for (ReceivedMessage message : receivedList) {
        ackIds.add(message.getAckId());
        Map<String, String> attributesMap = message.getMessage().getAttributesMap();
        out.addMessageLatency(
            Integer.parseInt(attributesMap.get("clientId")),
            Integer.parseInt(attributesMap.get("sequenceNumber")),
            System.currentTimeMillis() - Long.parseLong(attributesMap.get("sendTime")));
      }
      return out;
    };
    return Futures.transform(responseFuture, callback);
  }

  private synchronized SubscriberGrpc.SubscriberFutureStub getStub() {
    SubscriberGrpc.SubscriberFutureStub stub = stubs[currentStubIdx];
    currentStubIdx = (currentStubIdx + 1) % stubs.length;
    return stub;
  }
}

File: load-test-framework/python_src/clients/cps_subscriber_task.py
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from clients.task import Task, Worker, SubtaskWorker
from google.cloud.pubsub_v1.subscriber import Client
from google.cloud.pubsub_v1.subscriber.message import Message
import time
from clients.metrics_tracker import MessageAndDuration
import grpc
from concurrent import futures

from proto_dir.loadtest_pb2 import StartRequest
from proto_dir.loadtest_pb2_grpc import add_LoadtestWorkerServicer_to_server
from clients.loadtest_worker_servicer import LoadtestWorkerServicer
import sys


class SubscriberSubtaskWorker(SubtaskWorker):
    def run_worker(self, request: StartRequest):
        Worker.print_flush("started subscriber")
        subscription = "projects/" + request.project + "/subscriptions/" + request.pubsub_options.subscription
        client = Client()
        client.subscribe(subscription, self._on_receive).result()

    def _on_receive(self, message: Message):
        recv_time = int(time.time() * 1000)
        latency_ms = recv_time - int(message.attributes["sendTime"])
        pub_id = int(message.attributes["clientId"])
        sequence_number = int(message.attributes["sequenceNumber"])
        out = MessageAndDuration(pub_id, sequence_number, latency_ms)
        self.metrics_tracker.put(out)
        message.ack()


class CPSSubscriberWorker(Worker):
    def __init__(self):
        super().__init__(SubscriberSubtaskWorker())


class CPSSubscriberTask(Task):
    @staticmethod
    def get_worker() -> Worker:
        return CPSSubscriberWorker()


if __name__ == "__main__":
    port = "5000"
    for arg in sys.argv:
        if arg.startswith("--port="):
            port = arg.split("=")[1]
    task = CPSSubscriberTask()
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=2))
    add_LoadtestWorkerServicer_to_server(LoadtestWorkerServicer(task), server)
    address = '[::]:' + port
    server.add_insecure_port(address)
    server.start()
    print('subscriber server started at ' + address)
    while True:
        time.sleep(1)
